Deferred/Deferred.swift

```diff
@@ -6,39 +6,131 @@
 //  Copyright © 2014-2015 Big Nerd Ranch. Licensed under MIT.
 //
 
-// This cannot be a class var, new storage would be created for every
-// specialization. It also could not be used as a default argument as it is now.
-private var DeferredDefaultQueue: dispatch_queue_t {
-    return dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)
+import Dispatch
+
+private final class Storage<T> {
+
+    let value: T
+
+    init(_ value: T) {
+        self.value = value
+    }
+
+}
+
+// Atomic compare-and-swap, but safe for owned (retaining) pointers:
+//  - ObjC: "MyObject *__strong *"
+//  - Swift: "UnsafeMutablePointer<MyObject>"
+// If the swap is made, the new value is retained by its owning pointer.
+// If the swap is not made, the new value is not retained.
+private func compareAndSwap<T: AnyObject>(old old: T?, new: T?, to toPtr: UnsafeMutablePointer<T?>) -> Bool {
+    let oldRef = old.map(Unmanaged.passUnretained)
+    let newRef = new.map(Unmanaged.passRetained)
+    let oldPtr = oldRef?.toOpaque() ?? nil
+    let newPtr = newRef?.toOpaque() ?? nil
+    if OSAtomicCompareAndSwapPtr(UnsafeMutablePointer(oldPtr), UnsafeMutablePointer(newPtr), UnsafeMutablePointer(toPtr)) {
+        oldRef?.release()
+        return true
+    } else {
+        newRef?.release()
+        return false
+    }
+}
+
+// Raw Deferred storage. Using `ManagedBuffer` has advantages over a custom class:
+//  - The side-table data is efficiently stored in tail-allocated buffer space.
+//  - The Element buffer has a stable pointer when locked to a single element.
+//  - Better holdsUniqueReference support allows for future optimization.
+private final class DeferredBuffer<Value, SideData>: ManagedBuffer<SideData, Storage<Value>?> {
+    
+    static func create(sideData: () -> SideData) -> DeferredBuffer<Value, SideData> {
+        return create(1, initialValue: { _ in sideData() }) as! DeferredBuffer<Value, SideData>
+    }
+    
+    deinit {
+        // super's init automatically destroys the Value
```

Comment:
I think the point was that the comment says "super's *init* automatically destroys the value" vs "super's *deinit"…".